client/sniffer/src/main/java/org/elasticsearch/client/sniff/Sniffer.java

```diff
@@ -51,101 +53,167 @@
-                if (scheduledExecutorService.awaitTermination(1000, TimeUnit.MILLISECONDS)) {
-                    return;
-                }
-                scheduledExecutorService.shutdownNow();
-            } catch (InterruptedException e) {
-                Thread.currentThread().interrupt();
+                sniff();
+            } catch (Exception e) {
+                logger.error("error while sniffing nodes", e);
+            } finally {
+                Task task = new Task(sniffIntervalMillis);
+                Future<?> future = scheduler.schedule(task, nextTaskDelay);
+                //tasks are run by a single threaded executor, so swapping is safe with a simple volatile variable
+                ScheduledTask previousTask = nextScheduledTask;
+                nextScheduledTask = new ScheduledTask(task, future);
+                assert initialized.get() == false ||
+                        previousTask.task.isSkipped() || previousTask.task.hasStarted() : "task that we are replacing is neither " +
+                        "cancelled nor has it ever started";
             }
         }
+
+        /**
+         * Returns true if the task has started, false in case it didn't start (yet?) or it was skipped
+         */
+        boolean hasStarted() {
+            return taskState.get() == TaskState.STARTED;
+        }
+
+        /**
+         * Sets this task to be skipped. Returns true if the task will be skipped, false if the task has already started.
+         */
+        boolean skip() {
+            /*
+             Threads may still get run although future#cancel returns true. We make sure that a task is either cancelled (or skipped),
+             or entirely run. In the odd case that future#cancel returns true and the thread still runs, the task won't do anything.
+             In case future#cancel returns true but the task has already started, this state change will not succeed hence this method
+             returns false and the task will normally run.
+             */
+            return taskState.compareAndSet(TaskState.WAITING, TaskState.SKIPPED);
+        }
+
+        /**
+         * Returns true if the task was set to be skipped before it was started
+         */
+        boolean isSkipped() {
+            return taskState.get() == TaskState.SKIPPED;
+        }
+    }
+
+    static final class ScheduledTask {
+        final Task task;
+        final Future<?> future;
+
+        ScheduledTask(Task task, Future<?> future) {
+            this.task = task;
+            this.future = future;
+        }
+
+        /**
+         * Cancels this task. Returns true if the task has been successfully cancelled, meaning it won't be executed
+         * or if it is its execution won't have any effect. Returns false if the task cannot be cancelled (possibly it was
+         * already cancelled or already completed).
+         */
+        boolean cancelOrSKip() {
+            return future.cancel(false) && task.skip();
```

Comment:
I'd do
```
task.cancel(false);
return task.skip();
```

I don't think it is worth relying on `cancel` returning `true` in the case when we want to cancel. Maybe I'm overly paranoid but I don't trust it.

Reply:
And I might rename this method to `skip`.

Reply:
I agree! I have tests that verify this and I ran them many times but I am playing with fire here. After all, if cancel returns false whenever a task has already run, its corresponding state change will also fail. With this change all tests are still green so I think it's a good one to stay on the safe side.